### ISSUE TITLE
How to define timestamps in Carbon Black Cloud Syslog Connector 2.0

### ISSUE THREAD

**Issue by Kenshi-Katoh** [carbonblack/cbc-syslog#92]
### Describe the bug

Could you please help me with the timestamp of the logs of Carbon Black Cloud Syslog Connector 2.0.3?
Below is the log of vRLI, which is the forwarding destination of Carbon Black Cloud Syslog Connector.

"2024-07-24T04:45:02.885907Z localhost CEF:1|CarbonBlack|CBCSyslog|2.0.3|Audit Logs|Logged in successfully through Google Authenticator|1|rt=1721796131588 dvchost=xxxxx.co.jp duser=[xxxxx@xxxxx.co.jp](mailto:xxxxx@xxxxx.co.jp) dvc=xx.xx.xx.xx cs4Label=Event_ID cs4=xxxxxxxxxx"

"rt=" is the actual time when the log was output, but currently it is "rt=1721796131588",
which appears to be in UNIX timestamp format.
When converted to date and time, it becomes 2024/07/24 13:42:11, which matches the time when the log was actually output.

Please tell me how to output the "rt=" part in datetime_legacy (= Uses current time with format e.g. Jan 18 11:07:53) format that can be defined in the syslog header.

For example, in 1.3.1, the output is rt="Jun 07 2022 01:04:47".
I have attached the current configuration file so could you please check how to get the same output?

### Reproduction steps

1.Carbon Black Cloud Syslog Connector 2.0.3
2.Attached config file
3.(For example) rt=1721796131588
...

### Expected behavior

rt="Jun 07 2022 01:04:47"    (JST)

**Comment by Kenshi-Katoh** [comment]
[my-config.toml_20240725.txt](https://github.com/user-attachments/files/16956684/my-config.toml_20240725.txt)

**Comment by oono006** [comment]
This is an additional confirmation.
Would the following modifications result in the expected behavior?
Fixed parts are in bold and italicized.

### Before modification
[audit_logs_template]
template = "{{datetime_legacy}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|Audit Logs|{{description}}|1|{{extension}}"
type_field = ""
time_format = "%b %d %Y %H:%m:%S"
time_fields = ["eventTime"]

[audit_logs_template.extension]
default = "rt={{eventTime}}\tdvchost={{orgName}}\tduser={{loginName}}\tdvc={{clientIp}}\tcs4Label=Event_ID\tcs4={{eventId}}"

### After modification
[audit_logs_template]
template = "{{datetime_legacy}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|Audit Logs|{{description}}|1|{{extension}}"
type_field = ""
time_format = "%b %d %Y %H:%m:%S"
time_fields = ["eventTime"]

[audit_logs_template.extension]
***time_format = "%b %d %Y %H:%m:%S"***
***time_fields = ["eventTime"]***
default = "rt={{eventTime}}\tdvchost={{orgName}}\tduser={{loginName}}\tdvc={{clientIp}}\tcs4Label=Event_ID\tcs4={{eventId}}"

**Comment by avanbrunt-cb** [comment]
See the following section in the readme https://github.com/carbonblack/cbc-syslog?tab=readme-ov-file#creating-a-custom-message-with-templates

time_format and time_fields are non needed in the `[audit_logs_template.extension]`

Here is the example in the readme
```
 [alerts_template]
 template = "{{datetime_utc}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|{{reason_code}}|{{reason}}|{{severity}}|{{extension}}"
 type_field = "type"
 time_format = "%b %d %Y %H:%m:%S"
 time_fields = ["backend_timestamp"]

 [alerts_template.extension]
 default = "cat={{type}}\tact={{sensor_action}}\toutcome={{run_state}}"
 CB_ANALYTICS = "cat={{type}}\tact={{sensor_action}}\toutcome={{run_state}}\tframeworkName=MITRE_ATT&CK\tthreatAttackID={{attack_tactic}}:{{attack_technique}}"

 [audit_logs_template]
 template = "{{datetime_utc}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|Audit Logs|{{description}}|1|{{extension}}"
 time_format = "%b %d %Y %H:%m:%S"
 time_fields = ["eventTime"]

 [audit_logs_template.extension]
 default = "rt={{eventTime}}\tdvchost={{orgName}}\tduser={{loginName}}\tdvc={{clientIp}}\tcs4Label=Event_ID\tcs4={{eventId}}"
 ```

**Comment by avanbrunt-cb** [comment]
Note our timestamps are in UTC and are not convertible to a different timezone at this time. That will need to be handled at the destination

**Comment by oono006** [comment]
@avanbrunt-cb 

Thank you for contacting us.
Please allow me to confirm one additional point.

> Note our timestamps are in UTC and are not convertible to a different timezone at this time. That will need to be handled at the destination

As mentioned above, the version for which this phenomenon (timestamp in UTC) has been confirmed is 2.0.3.
We are aware that this has not occurred in 1.3.1.
From which version is the event occurring?

**Comment by avanbrunt-cb** [comment]
We do not support CBC Syslog 1.3.1 any more and the only supported versions are CBC Syslog 2.x.

The new APIs behind the CBC Syslog 2.x version only support UTC timestamps and can be configured as you desire by providing a compatible strftime string in the `time_format` option.

From the readme:

`time_format` and `time_fields` provides you the ability to customize the way the timestamps are formatted and which fields to modify. This utilizes python strftime formatting, for more information on strftime formats see https://strftime.org/

**Comment by oono006** [comment]
@avanbrunt-cb 

Thank you for your response.

> The new APIs behind the CBC Syslog 2.x version only support UTC timestamps and can be configured as you desire by providing a compatible strftime string in the time_format option.

We are aware that the CBC Syslog 2.x version only supports UTC timestamps.
We have confirmed this for https://strftime.org/.
From the following configuration, I expected the output to be rt=“Jun 07 2022 01:04:47” as in CBC Syslog 1.3.1.

> time_format = "%b %d %Y %H:%m:%S"
> time_fields = ["eventTime"]

Could you please respond to the following two points?

1. I am afraid I am repeating myself,
Please let me know how to configure it to output date and time type output like rt=“Jun 07 2022 01:04:47”.

2. Please let us know if the CBC Syslog configuration does not support date and time type output.
The syslog is sent to vRealize Log Insight.
> That will need to be handled at the destination

**Comment by avanbrunt-cb** [comment]
Yes `%b %d %Y %H:%m:%S` corresponds to the timestamp format of `Jun 07 2022 01:04:47` 

The following example will set the timestamp as you expect using the CEF syslog format where `rt=“Jun 07 2022 01:04:47”` is included in the extension

```
[audit_logs_template]
 template = "{{datetime_utc}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|Audit Logs|{{description}}|1|{{extension}}"
 time_format = "%b %d %Y %H:%m:%S"
 time_fields = ["eventTime"]

 [audit_logs_template.extension]
 default = "rt={{eventTime}}\tdvchost={{orgName}}\tduser={{loginName}}\tdvc={{clientIp}}\tcs4Label=Event_ID\tcs4={{eventId}}"
 ```
 
 If you want to try a few options then use `file` as the `output_type` with `template` as the `output_format`

**Comment by oono006** [comment]
@avanbrunt-cb 
Thank you for your response.

> Yes `%b %d %Y %H:%m:%S` corresponds to the timestamp format of `Jun 07 2022 01:04:47`
> 
> The following example will set the timestamp as you expect using the CEF syslog format where `rt=“Jun 07 2022 01:04:47”` is included in the extension
> 
> ```
> [audit_logs_template]
>  template = "{{datetime_utc}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|Audit Logs|{{description}}|1|{{extension}}"
>  time_format = "%b %d %Y %H:%m:%S"
>  time_fields = ["eventTime"]
> 
>  [audit_logs_template.extension]
>  default = "rt={{eventTime}}\tdvchost={{orgName}}\tduser={{loginName}}\tdvc={{clientIp}}\tcs4Label=Event_ID\tcs4={{eventId}}"
> ```
> 
> If you want to try a few options then use `file` as the `output_type` with `template` as the `output_format`

It seems to be the same as the config I have previously drafted, but it does not work as expected.
Is the following wording unnecessary?

**_type_field = “”_**

Just to be sure, here is the config.
https://github.com/carbonblack/cbc-syslog/issues/92#issuecomment-2342440804

**Comment by avanbrunt-cb** [comment]
I'm not sure where you are getting `type_field = “”`. You use the following to specify which fields you want modified to the specified format `time_fields = ["eventTime"]` then in the extensions you include that field value `default = "rt={{eventTime}}`

Can you provide your current configuration and a sample output?

**Comment by oono006** [comment]
@avanbrunt-cb 

Thank you for your response.
The current configuration and output as drafted here is as follows.
https://github.com/carbonblack/cbc-syslog/issues/92#issue-2518251769

"rt=" is the actual time when the log was output, but currently it is "rt=1721796131588",
which appears to be in UNIX timestamp format.
When converted to date and time, it becomes 2024/07/24 13:42:11, which matches the time when the log was actually output.

I would like to output the data in a date/time type like 
"rt=Jun 07 2022 01:04:47” instead of the UNIX timestamp format.



### Configuration
```
[audit_logs_template]
template = "{{datetime_utc}} localhost CEF:1|{{vendor}}|{{product}}|{{product_version}}|Audit Logs|{{description}}|1|{{extension}}"
type_field = ""
time_format = "%b %d %Y %H:%m:%S"
time_fields = ["eventTime"]

[audit_logs_template.extension]
default = "rt={{eventTime}}\tdvchost={{orgName}}\tduser={{loginName}}\tdvc={{clientIp}}\tcs4Label=Event_ID\tcs4={{eventId}}"
```
### Output
`"2024-07-24T04:45:02.885907Z localhost CEF:1|CarbonBlack|CBCSyslog|2.0.3|Audit Logs|Logged in successfully through Google Authenticator|1|rt=1721796131588 dvchost=xxxxx.co.jp duser=[xxxxx@xxxxx.co.jp](mailto:xxxxx@xxxxx.co.jp) dvc=xx.xx.xx.xx cs4Label=Event_ID cs4=xxxxxxxxxx"`

**Comment by avanbrunt-cb** [comment]
Ah I missed the issue. Sorry for the confusion. It appears the timestamp we are receiving from the backend is in milliseconds which causes an issue with the transform. I will get a fix delivered tomorrow

**Comment by oono006** [comment]
@avanbrunt-cb 
Thank you for your response.

We understand that you can provide us with a corrected version.
Thank you very much for your cooperation.

**Comment by avanbrunt-cb** [comment]
Here is a branch if you want to test https://github.com/carbonblack/cbc-syslog/tree/fix-epoch-milliseconds I'll try to get a new release cut tomorrow

**Comment by oono006** [comment]
@avanbrunt-cb 
Thank you very much for your prompt response.
We look forward to working with you.

**Comment by oono006** [comment]
@avanbrunt-cb 
Sorry for the continued inquiry.
We have heard that you can fix the problem, but which of the following would you suggest?

1. Upgrade of CBC Syslog Connector
2. Modification of the CEF template (https://github.com/carbonblack/cbc-syslog)


Please confirm the above.

**Comment by avanbrunt-cb** [comment]
This has been fixed with CBC Syslog 2.0.6. Please upgrade to the latest version

**Comment by oono006** [comment]
@avanbrunt-cb 
We apologize for the delay in contacting you.

Thank you very much for your response to this matter.
It was very helpful.
Please close this issue.


Thank you very much for your continued support.
